daemon/containerd/image.go

```diff
@@ -2,14 +2,241 @@ package containerd
 
 import (
 	"context"
-	"errors"
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"strconv"
+	"sync/atomic"
+	"time"
 
+	"github.com/containerd/containerd/content"
+	cerrdefs "github.com/containerd/containerd/errdefs"
+	containerdimages "github.com/containerd/containerd/images"
+	"github.com/containerd/containerd/platforms"
+	"github.com/docker/distribution/reference"
+	containertypes "github.com/docker/docker/api/types/container"
 	imagetype "github.com/docker/docker/api/types/image"
+	"github.com/docker/docker/daemon/images"
 	"github.com/docker/docker/errdefs"
 	"github.com/docker/docker/image"
+	"github.com/docker/docker/layer"
+	"github.com/docker/go-connections/nat"
+	"github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+	"golang.org/x/sync/semaphore"
 )
 
+var truncatedID = regexp.MustCompile(`^([a-f0-9]{4,64})$`)
+
 // GetImage returns an image corresponding to the image referred to by refOrID.
-func (i *ImageService) GetImage(ctx context.Context, refOrID string, options imagetype.GetImageOpts) (retImg *image.Image, retErr error) {
-	return nil, errdefs.NotImplemented(errors.New("not implemented"))
+func (i *ImageService) GetImage(ctx context.Context, refOrID string, options imagetype.GetImageOpts) (*image.Image, error) {
+	desc, err := i.resolveDescriptor(ctx, refOrID)
+	if err != nil {
+		return nil, err
+	}
+
+	if options.Platform != nil {
+		cs := i.client.ContentStore()
+		imgPlatforms, err := containerdimages.Platforms(ctx, cs, desc)
+		if err != nil {
+			return nil, err
+		}
+
+		platformFound := false
+		comparer := platforms.Only(*options.Platform)
+		for _, p := range imgPlatforms {
+			if comparer.Match(p) {
+				platformFound = true
+				break
+			}
+		}
+		if !platformFound {
+			return nil, errdefs.NotFound(errors.Errorf("%s: platform %s not supported by image", refOrID, platforms.Format(*options.Platform)))
+		}
+	}
+
+	cs := i.client.ContentStore()
+
+	platform := platforms.Default()
```

Comment:
Hm, this won't work, if we have image for the non-default platform only (like if we only do `docker pull --platform linux/arm64 alpine` on non-arm64 host).
Is this expected?

Reply:
I think it makes sense that when no `--platform` flag is given then we consider this as the user asking for the current platform

Reply:
That bugs me - if the user doesn't set the `--platform` then we don't know what he's asking for, consider this:
```bash
$ docker pull --platform linux/arm64 alpine
$ docker inspect alpine
alpine: platform linux/amd64 not supported by image
```

seems a bit harsh, especially for doing an `inspect` which should allow me to get information about an image. The user might not remember which platform the image was pulled for and currently there's no way for him to find out that.

Reply:
What should we show in this case then?
```console
$ docker pull --platform linux/arm64 alpine
$ docker pull --platform linux/s390x alpine
$ docker inspect alpine
```

Reply:
The "old" behavior would be to pick the first platform that's found (but of course that would always be the _only_ platform). Ideally we would show all platforms for the image, but that brings us to the discussion in https://github.com/moby/moby/issues/44582. If we _must_ show a single architecture, but multiple architectures are present, it could follow the same method of selecting as a `docker pull` would do (basically, use a platform-matcher?)

Reply:
Consider the API currently shows whatever image is currently there regardless of platform.

We could error out with like `ambiguous image reference: multiple platforms are available, you must specify a platform: available platforms <platform1>,<platform2>`. It would be nice, though, if the API's json error could list those platforms so clients can parse it nicely (without having to parse the error string) or pick the first one... I'm inclined to error in case of ambiguity here, though.

Alternatively... we could update the API to return a list of objects
The docker CLI at least already outputs a json array.
There wouldn't be much too change to support the updated API.

... or change it to return the image index (with the appropriate media type on the response headers).
This does mean larger changes would be needed on the client.

Reply:
On the last point... we could gate this by accept header.

Reply:
Talking with @thaJeztah and we decided to go with a [AllPlatformsWithPreference](https://github.com/moby/moby/pull/43818/files#diff-1882c3a96e0a351a1471f851024987ed299b44977c119020244e91c54dca3a10R14) that returns the preferred platform first and then the others in order that they are defined by containerd.

Reply:
^^ That approach is the "first step" for this PR, to keep the behavior (mostly) consistent with what we have pre-containers, and to get us going, but we should indeed continue the discussion on UX (which could be return multiple platforms if there's multiple variations present)

Reply:
Definitely, the discussion is only getting started. Way back when we talked about snapshotters the decided way to go was to make everything as close as possible to the existing and _after_ that we can shape the things in different ways